exec/java-exec/src/main/java/org/apache/drill/exec/physical/resultSet/project/RequestedColumn.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.exec.physical.resultSet.project;
+
+/**
+ * Plan-time properties of a requested column.Represents
+ * a consolidated view of the set of references to a column.
+ * For example, the project list might contain:<br>
+ * <tt>SELECT columns[4], columns[8]</tt><br>
+ * <tt>SELECT a.b, a.c</tt><br>
+ * <tt>SELECT columns, columns[1]</tt><br>
+ * <tt>SELECT a, a.b</tt><br>
+ * In each case, the same column is referenced in different
+ * forms which are consolidated into this abstraction.
+ * <p>
+ * The resulting information is a "pattern": a form of reference
+ * which which a concrete type can be compatible or not. The project
+ * list does not contain sufficient information to definitively pick
+ * a type; it only excludes certain types.
+ * <p>
+ * Depending on the syntax, we can infer if a column must
+ * be an array or map. This is definitive: though we know that
+ * columns of the form above must be an array or a map,
+ * we cannot know if a simple column reference might refer
+ * to an array or map.
+ *
+ * <h4>Compatibility Rules<h4>
+ *
+ * The pattern given by projection is consistent with certain concrete types
+ * as follows. + means any number of additional qualifiers.
+ * <p>
+ * <table>
+ * <tr><th>Type</th><th>Consistent with</th></tr>
+ * <tr><td>Non-repeated MAP</td>
+ *     <td>{@code a+} {@code a.b+}</td></tr>
+ * <tr><td>Repeated MAP</td>
+ *     <td>{@code a+} {@code a.b+} {@code a[n].b+}</td>></tr>
+ * <tr><td>Non-repeated Scalar</td>
+ *     <td>{@code a}</td></tr>
+ * <tr><td>Repeated Scalar</td>
+ *     <td>{@code a} {@code a[n]}</td></tr>
+ * <tr><td>Non-repeated DICT</td>
+ *     <td>{@code a} {@code a['key']}</td></tr>
+ * <tr><td>Repeated DICT</td>
+ *     <td>{@code a} {@code a[n]} {@code a['key']} {@code a[n]['key']}</td></tr>
```

Comment:
I am not sure whether `a['key']` pattern should be supported for `Repeated DICT` columns - maybe an UDF, e.g. `ARRAY<V> dict_values(DICT<K,V>)`, suits better for this purpose? As it provides clearer intentions and avoids mistakes.
What do you think?

Reply:
Agreed. Changed the wording to suggest the mappings are conceptual. I *think* we allow `m.a` for `MAP` arrays, the Project operator(?) will pull out that one column as a top-level array. Conceptually, we could do the same for `DICT`. Implementation-wise, your suggestion might be more practical.

Reply:
Checked whether `m.a` is supported for `MAP` arrays: it looks like this is not supported in Drill.
For a json file `file.json`
```
{"sa": [{"a": 1}, {"a": 2}, {"a": 3}]}
{"sa": [{"a": 1}]}
```
following query ``select t.sa.a kv from dfs.`file.json` t`` produces two rows with `null` value each. (Should have returned an error instead?)

In case when types are known during planning, e.g. in case when querying Hive table, there is following validation: `VALIDATION ERROR: From line 1, column 27 to line 1, column 28: Cannot apply 'ITEM' to arguments of type 'ITEM(<RECORDTYPE(INTEGER SN, RECORDTYPE(INTEGER TN, VARCHAR(65535) TS) ARRAY SA) ARRAY>, <CHAR(2)>)'. Supported form(s): <ARRAY>[<INTEGER>]
<MAP>[<VALUE>]`
(used following test in `TestHiveStructs.java`:
```
@Test
  public void strWithArr2ByIdxP0111() throws Exception {
    HiveTestUtilities.assertNativeScanUsed(queryBuilder(), "struct_tbl_p");
    testBuilder()
        .sqlQuery("SELECT rid, t.str_wa_2.fa.sn p0 FROM hive.struct_tbl_p t")
        .unOrdered()
        .baselineColumns("rid", "p0")
        .expectsEmptyResultSet()
        .go();
  }
```
)

However, such behavior is present in Hive, but for repeated (Drill's) `MAP` only (but not for repeated `DICT`), IIRC.

Reply:
@KazydubB , thanks for checking. It seems that the project operator is making up a null column since it does not know how to do the actual projection. I'll see how I can incorporate this additional restriction into the code here.